### ISSUE TITLE
[ML-02] Fine-tune FaceNet (InceptionResnetV1) and export facenet_best.onnx

### ISSUE THREAD

**Issue by victorjudysen** [victorjudysen/ai-exam-proctoring-system#84]
## Context

AI-Based Online Exam Proctoring System — ML Training (Google Colab). Depends on [ML-01] (datasets ready). This is the most critical ML deliverable — the ONNX model is consumed by Kweka's AI Service ([AI-03]).

**Hard deadline: model must be shared with Kweka via Google Drive before AI-03 is tested end-to-end.**

### `ml-training/02_facenet_training.ipynb`

**Cell 1 — Load model:**
```python
from facenet_pytorch import InceptionResnetV1, MTCNN
import torch, torch.nn as nn

device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
model = InceptionResnetV1(pretrained='vggface2', classify=False).to(device)
model.train()
```

**Cell 2 — Dataset loader:**
```python
from torch.utils.data import DataLoader, Dataset
from PIL import Image
import os

class FaceDataset(Dataset):
    """Loads MTCNN-aligned faces. Filters to identities with >= 2 images for triplet loss."""
    def __init__(self, root_dir, mtcnn, min_images=2):
        ...
    def __len__(self): ...
    def __getitem__(self, idx): ...   # returns (image_tensor, label)

# 70/15/15 split by identity
train_dataset = FaceDataset('/content/drive/MyDrive/fyp-ai/datasets/lfw/lfw', mtcnn)
```

**Cell 3 — Training with Triplet Loss:**
```python
from torch.optim import Adam

criterion = nn.TripletMarginLoss(margin=0.2)
optimizer = Adam(model.parameters(), lr=1e-4)

for epoch in range(20):
    for anchor, positive, negative in triplet_loader:
        anchor, positive, negative = anchor.to(device), positive.to(device), negative.to(device)
        emb_a = model(anchor)
        emb_p = model(positive)
        emb_n = model(negative)
        loss = criterion(emb_a, emb_p, emb_n)
        optimizer.zero_grad()
        loss.backward()
        optimizer.step()
    print(f"Epoch {epoch+1}/20 — Loss: {loss.item():.4f}")
    # Save checkpoint every 5 epochs
    if (epoch+1) % 5 == 0:
        torch.save(model.state_dict(), f'/content/drive/MyDrive/fyp-ai/checkpoints/facenet_epoch{epoch+1}.pt')
```

**Cell 4 — Evaluation (FAR/FRR on LFW test split):**
```python
def evaluate(model, test_pairs):
    """
    Compute accuracy, FAR, FRR.
    Target: accuracy >= 90%, FAR < 5%, FRR < 10%
    """
    model.eval()
    ...
```

**Cell 5 — ONNX Export:**
```python
import torch.onnx

model.eval()
dummy_input = torch.randn(1, 3, 160, 160).to(device)
torch.onnx.export(
    model,
    dummy_input,
    '/content/drive/MyDrive/fyp-ai/exports/facenet_best.onnx',
    input_names=['input'],
    output_names=['embedding'],
    opset_version=14,
    dynamic_axes={'input': {0: 'batch_size'}, 'embedding': {0: 'batch_size'}}
)
print("ONNX model exported successfully")
```

**Cell 6 — Verify ONNX export:**
```python
import onnxruntime as ort, numpy as np
sess = ort.InferenceSession('/content/drive/MyDrive/fyp-ai/exports/facenet_best.onnx')
dummy = np.random.randn(1, 3, 160, 160).astype(np.float32)
output = sess.run(['embedding'], {'input': dummy})
print(f"Output shape: {output[0].shape}")   # Expected: (1, 512)
```

## Input/Output Contract (for Kweka's AI-03)

```
Input:  float32 tensor, shape (1, 3, 160, 160)
        Normalised: mean=[0.5, 0.5, 0.5], std=[0.5, 0.5, 0.5]
        Face must be MTCNN-aligned and cropped to 160x160

Output: float32 tensor, shape (1, 512)
        L2-normalised embedding vector
```

## Sharing the model with Kweka

After export:
1. Right-click `facenet_best.onnx` in Google Drive
2. Share → "Anyone with the link can view"
3. Post the share link in the team WhatsApp group
4. Kweka will download it to `ai-service/models/facenet_best.onnx`

## Quality Targets

| Metric | Target |
|--------|--------|
| Accuracy (LFW test) | ≥ 90% |
| False Acceptance Rate (FAR) | < 5% |
| False Rejection Rate (FRR) | < 10% |
| Output shape | (1, 512) float32 |
| File size | < 100MB |

## Branch

Upload notebook to repo: `ml-training/02_facenet_training.ipynb`
Commit: `git commit -m "ml: add FaceNet training and ONNX export notebook"`
